src/main/java/org/jenkinsci/plugins/workflow/graphanalysis/ForkScanner.java

```diff
@@ -244,44 +243,20 @@ public boolean isLeaf() {
         }
 
         /**
-         * We have discovered a forking node intersecting our FlowSegment in the middle or meeting at the end
-         * Now we need to split the flow, or pull out the fork point and make both branches follow it
+         * We have discovered a parallel step start node at the end of our segment, so we create a fork for that step and point it to this segment.
          * @param nodeMapping Mapping of BlockStartNodes to flowpieces (forks or segments)
-         * @param joinPoint Node where the branches intersect/meet (fork point)
-         * @param joiningBranch Flow piece that is joining this
-         * @throws IllegalStateException When you try to split a segment on a node that it doesn't contain, or invalid graph structure
+         * @param joinPoint The parallel step start node
+         * @throws IllegalStateException When you try to split a segment in an invalid way
          * @return Recreated fork
          */
-        Fork split(@NonNull HashMap<FlowNode, FlowPiece> nodeMapping, @NonNull BlockStartNode joinPoint, @NonNull FlowPiece joiningBranch) {
-            int index = visited.lastIndexOf(joinPoint);  // Fork will be closer to end, so this is better than indexOf
-            Fork newFork = new Fork(joinPoint);
-
-            if (index < 0) {
-                throw new IllegalStateException("Tried to split a segment where the node doesn't exist in this segment");
-            } else if (index == this.visited.size()-1) { // We forked just off the most recent node
-                newFork.following.add(this);
-                newFork.following.add(joiningBranch);
-                this.visited.remove(index);
-            } else if (index == 0) {
-                throw new IllegalStateException("We have a cyclic graph or heads that are not separate branches!");
-            } else { // Splitting at some midpoint within the segment, everything before becomes part of the following
-                // Execute the split: create a new fork at the fork point, and shuffle the part of the flow after it
-                //   to a new segment and add that to the fork.
-
-                FlowSegment newSegment = new FlowSegment();
-                newSegment.after = this.after;
-                newSegment.visited.addAll(this.visited.subList(0, index));
-                newFork.following.add(newSegment);
-                newFork.following.add(joiningBranch);
-                this.after = newFork;
-                this.isLeaf = false;
-
-                // Remove the part before the fork point
-                this.visited.subList(0, index+1).clear();
-                for (FlowNode n : newSegment.visited) {
-                    nodeMapping.put(n, newSegment);
-                }
```

Comment:
This logic previously had branches to account for being called in various ways. I changed the way it is called making all but one of the branches unreachable.